src/Workspaces/Core/Portable/Classification/SyntaxClassification/SyntacticChangeRangeComputer.cs

```diff
@@ -0,0 +1,249 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Classification
+{
+    /// <summary>
+    /// Computes a syntactic text change range that determines the range of a document that was changed by an edit. The
+    /// portions outside this change range are guaranteed to be syntactically identical (see <see
+    /// cref="SyntaxNode.IsIncrementallyIdenticalTo"/>).  This algorithm is intended to be <em>fast</em>.  It is
+    /// technically linear in the number of nodes and tokens that may need to examined.  However, in practice, it should
+    /// operate in sub-linear time as it will bail the moment tokens don't match, and it's able to skip over matching
+    /// nodes fully without examining the contents of those nodes.  This is intended for consumers that want a
+    /// reasonably accurate change range computer, but do not want to spend an inordinate amount of time getting the
+    /// most accurate and minimal result possible.
+    /// </summary>
+    /// <remarks>
+    /// This computation is not guaranteed to be minimal.  It may return a range that includes parts that are unchanged.
+    /// This means it is also legal for the change range to just specify the entire file was changed. The quality of
+    /// results will depend on how well the parsers did with incremental parsing, and how much time is given to do the
+    /// comparison.  In practice, for large files (i.e. 15kloc) with standard types of edits, this generally returns
+    /// results in around 50-100 usecs on a i7 3GHz desktop.
+    /// <para>
+    /// This algorithm will respect the timeout provided to the best of abilities.  If any information has been computed
+    /// when the timeout elapses, it will be returned.
+    /// </para>
+    /// </remarks>
+    internal static class SyntacticChangeRangeComputer
+    {
+        private static readonly ObjectPool<Stack<SyntaxNodeOrToken>> s_pool = new(() => new());
+
+        public static async ValueTask<TextChangeRange?> ComputeSyntacticChangeRangeAsync(
+            Document oldDocument, Document newDocument, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            // If they're the same doc, there is no change.
+            if (oldDocument == newDocument)
+                return new TextChangeRange();
```

Comment:
```suggestion
                return new();
```

Tiny nit

Reply:
my preference is to use `new()` in places that are immediatley apparent.  So with `X x = new()` or `X Foo() => new()`.  Outside of that, i'm currently using sparingly :)